diff --git a/kernel/Kconfig.hz b/kernel/Kconfig.hz
index ce1435cb08b1..1abec209df38 100644
--- a/kernel/Kconfig.hz
+++ b/kernel/Kconfig.hz
@@ -24,6 +24,12 @@
 	  with lots of processors that may show reduced performance if
 	  too many timer interrupts are occurring.
 
+	config HZ_175
+		bool "175 HZ"
+	help
+	  175 Hz is a slightly better timer frequency for small VMs within
+	  cloud.
+
 	config HZ_250
 		bool "250 HZ"
 	help
@@ -31,6 +37,11 @@
 	 while also showing good interactive responsiveness even
 	 on SMP and NUMA systems. If you are going to be using NTSC video
 	 or multimedia, select 300Hz instead.
+
+	config HZ_275
+		bool "275 HZ"
+	help
+	 275 HZ is a better compromise for latency and throuput.
 
 	config HZ_300
 		bool "300 HZ"
@@ -51,7 +62,9 @@
 config HZ
 	int
 	default 100 if HZ_100
+	default 175 if HZ_175
 	default 250 if HZ_250
+	default 275 if HZ_275
 	default 300 if HZ_300
 	default 1000 if HZ_1000
 
