--- a/arch/arm64/boot/dts/rockchip/rk3588s-orangepi-5-pro.dts	2026-10-10 04:18:20.000000000 +0000
+++ b/arch/arm64/boot/dts/rockchip/rk3588s-orangepi-5-pro.dts	2026-10-10 04:20:01.273126658 +0000
@@ -69,7 +69,7 @@
 		#cooling-cells = <2>;
 		cooling-levels = <0 50 100 150 200 255>;
 		fan-supply = <&vcc5v0_sys>;
-		pwms = <&pwm2 0 20000000 0>;
+		pwms = <&pwm2 0 16666666 0>;
 	};
 
 	vcc3v3_eth: regulator-vcc3v3-eth {
