--- a/net/ipv4/tcp_bbr.c	2026-05-15 02:23:53.934546500 +0800
+++ b/net/ipv4/tcp_bbr.c	2026-05-15 20:28:22.832044900 +0800
@@ -132,74 +132,84 @@
 /* Window length of bw filter (in rounds): */
 static const int bbr_bw_rtts = CYCLE_LEN + 2;
 /* Window length of min_rtt filter (in sec): */
-static const u32 bbr_min_rtt_win_sec = 10;
+static const u32 bbr_min_rtt_win_sec = 15;
 /* Minimum time (in ms) spent at bbr_cwnd_min_target in BBR_PROBE_RTT mode: */
-static const u32 bbr_probe_rtt_mode_ms = 200;
+static const u32 bbr_probe_rtt_mode_ms = 140;
 /* Skip TSO below the following bandwidth (bits/sec): */
 static const int bbr_min_tso_rate = 1200000;
 
-/* Pace at ~1% below estimated bw, on average, to reduce queue at bottleneck.
+/* Pace at ~2% below estimated bw, on average, to reduce queue at bottleneck.
  * In order to help drive the network toward lower queues and low latency while
  * maintaining high utilization, the average pacing rate aims to be slightly
  * lower than the estimated bandwidth. This is an important aspect of the
  * design.
  */
-static const int bbr_pacing_margin_percent = 1;
+static const int bbr_pacing_margin_percent = 2;
 
 /* We use a high_gain value of 2/ln(2) because it's the smallest pacing gain
  * that will allow a smoothly increasing pacing rate that will double each RTT
  * and send the same number of packets per RTT that an un-paced, slow-starting
  * Reno or CUBIC flow would:
  */
-static const int bbr_high_gain  = BBR_UNIT * 2885 / 1000 + 1;
+static const int bbr_high_gain  = BBR_UNIT * 303 / 100 + 1;
 /* The pacing gain of 1/high_gain in BBR_DRAIN is calculated to typically drain
  * the queue created in BBR_STARTUP in a single round:
  */
-static const int bbr_drain_gain = BBR_UNIT * 1000 / 2885;
+static const int bbr_drain_gain = BBR_UNIT * 1000 / 2414;
 /* The gain for deriving steady-state cwnd tolerates delayed/stretched ACKs: */
-static const int bbr_cwnd_gain  = BBR_UNIT * 2;
+static const int bbr_cwnd_gain  = BBR_UNIT * 27 / 10;
+/* The gain for deriving startup cwnd: */
+static const int bbr_startup_cwnd_gain = BBR_UNIT * 5 / 2;
 /* The pacing_gain values for the PROBE_BW gain cycle, to discover/share bw: */
 static const int bbr_pacing_gain[] = {
-	BBR_UNIT * 5 / 4,	/* probe for more available bw */
-	BBR_UNIT * 3 / 4,	/* drain queue and/or yield bw to other flows */
-	BBR_UNIT, BBR_UNIT, BBR_UNIT,	/* cruise at 1.0*bw to utilize pipe, */
-	BBR_UNIT, BBR_UNIT, BBR_UNIT	/* without creating excess queue... */
+	BBR_UNIT * 3,	/* UP: probe for more available bw */
+	BBR_UNIT * 87 / 100,	/* DOWN: drain queue and/or yield bw */
+	BBR_UNIT * 11 / 10,	/* CRUISE: try to use pipe w/ some headroom */
+	BBR_UNIT * 5 / 4,	/* REFILL: refill pipe to estimated 100% */
+	BBR_UNIT * 2,
+	BBR_UNIT * 4 / 5,
+	BBR_UNIT * 51 / 50,
+	BBR_UNIT * 6 / 5,
 };
 /* Randomize the starting gain cycling phase over N phases: */
-static const u32 bbr_cycle_rand = 7;
+static const u32 bbr_cycle_rand = 2;
 
 /* Try to keep at least this many packets in flight, if things go smoothly. For
  * smooth functioning, a sliding window protocol ACKing every other packet
  * needs at least 4 packets in flight:
  */
-static const u32 bbr_cwnd_min_target = 4;
+static const u32 bbr_cwnd_min_target = 12;
+
+/* Avoid over pacing when app-limited in STARTUP, but still try to probe for more bandwidth.
+ */
+static const u32 bbr_limited_pacing_gain = BBR_UNIT * 6 / 5;
 
 /* To estimate if BBR_STARTUP mode (i.e. high_gain) has filled pipe... */
-/* If bw has increased significantly (1.25x), there may be more bw available: */
-static const u32 bbr_full_bw_thresh = BBR_UNIT * 5 / 4;
-/* But after 3 rounds w/o significant bw growth, estimate pipe is full: */
-static const u32 bbr_full_bw_cnt = 3;
+/* If bw has increased significantly (1.19x), there may be more bw available: */
+static const u32 bbr_full_bw_thresh = BBR_UNIT * 119 / 100;
+/* But after 4 rounds w/o significant bw growth, estimate pipe is full: */
+static const u32 bbr_full_bw_cnt = 4;
 
 /* "long-term" ("LT") bandwidth estimator parameters... */
 /* The minimum number of rounds in an LT bw sampling interval: */
 static const u32 bbr_lt_intvl_min_rtts = 4;
 /* If lost/delivered ratio > 20%, interval is "lossy" and we may be policed: */
-static const u32 bbr_lt_loss_thresh = 50;
-/* If 2 intervals have a bw ratio <= 1/8, their bw is "consistent": */
-static const u32 bbr_lt_bw_ratio = BBR_UNIT / 8;
-/* If 2 intervals have a bw diff <= 4 Kbit/sec their bw is "consistent": */
-static const u32 bbr_lt_bw_diff = 4000 / 8;
+static const u32 bbr_lt_loss_thresh = 42;
+/* If 2 intervals have a bw ratio <= 1/9, their bw is "consistent": */
+static const u32 bbr_lt_bw_ratio = BBR_UNIT / 9;
+/* If 2 intervals have a bw diff <= 7 Kbit/sec their bw is "consistent": */
+static const u32 bbr_lt_bw_diff = 7000 / 8;
 /* If we estimate we're policed, use lt_bw for this many round trips: */
-static const u32 bbr_lt_bw_max_rtts = 48;
+static const u32 bbr_lt_bw_max_rtts = 21;
 
 /* Gain factor for adding extra_acked to target cwnd: */
-static const int bbr_extra_acked_gain = BBR_UNIT;
+static const int bbr_extra_acked_gain = BBR_UNIT * 27 / 25;
 /* Window length of extra_acked window. */
 static const u32 bbr_extra_acked_win_rtts = 5;
 /* Max allowed val for ack_epoch_acked, after which sampling epoch is reset */
 static const u32 bbr_ack_epoch_acked_reset_thresh = 1U << 20;
 /* Time period for clamping cwnd increment due to ack aggregation */
-static const u32 bbr_extra_acked_max_us = 100 * 1000;
+static const u32 bbr_extra_acked_max_us = 118 * 1000;
 
 static void bbr_check_probe_rtt_done(struct sock *sk);
 
@@ -288,11 +298,17 @@
 	struct tcp_sock *tp = tcp_sk(sk);
 	struct bbr *bbr = inet_csk_ca(sk);
 	unsigned long rate = bbr_bw_to_pacing_rate(sk, bw, gain);
+	bool force_update = unlikely(bbr->mode == BBR_STARTUP && tp->app_limited);
 
 	if (unlikely(!bbr->has_seen_rtt && tp->srtt_us))
 		bbr_init_pacing_rate_from_rtt(sk);
-	if (bbr_full_bw_reached(sk) || rate > READ_ONCE(sk->sk_pacing_rate))
+	if (bbr_full_bw_reached(sk) || rate > READ_ONCE(sk->sk_pacing_rate) || force_update) {
+		if (unlikely(bbr->min_rtt_us <= 750)) {
+			gain = gain > BBR_UNIT ? min(gain, BBR_UNIT * 5 / 4) : gain;
+			rate = bbr_bw_to_pacing_rate(sk, bw, gain);
+		}
 		WRITE_ONCE(sk->sk_pacing_rate, rate);
+	}
 }
 
 /* override sysctl_tcp_min_tso_segs */
@@ -362,6 +378,7 @@
 	struct bbr *bbr = inet_csk_ca(sk);
 	u32 bdp;
 	u64 w;
+	u32 minrtt = bbr->min_rtt_us;
 
 	/* If we've never had a valid RTT sample, cap cwnd at the initial
 	 * default. This should only happen when the connection is not using TCP
@@ -369,16 +386,36 @@
 	 * ACKed so far. In this case, an RTO can cut cwnd to 1, in which
 	 * case we need to slow-start up toward something safe: TCP_INIT_CWND.
 	 */
-	if (unlikely(bbr->min_rtt_us == ~0U))	 /* no valid RTT samples yet? */
+	if (unlikely(minrtt == ~0U))	 /* no valid RTT samples yet? */
 		return TCP_INIT_CWND;  /* be safe: cap at default initial cwnd*/
 
-	w = (u64)bw * bbr->min_rtt_us;
+	w = (u64)bw * minrtt;
 
 	/* Apply a gain to the given value, remove the BW_SCALE shift, and
 	 * round the value up to avoid a negative feedback loop.
 	 */
 	bdp = (((w * gain) >> BBR_SCALE) + BW_UNIT - 1) / BW_UNIT;
 
+	if (minrtt >= 180 * USEC_PER_MSEC) {
+			bdp += 8;
+			bdp += bw >= 44738 ? 6 : 3;
+		} else if (minrtt >= 140 * USEC_PER_MSEC) {
+			bdp += 6;
+			bdp += bw >= 44738 ? 5 : 2;
+		} else if (minrtt >= 90 * USEC_PER_MSEC) {
+			bdp += 5;
+			bdp += bw >= 44738 ? 4 : 1;
+		} else if (minrtt >= 31 * USEC_PER_MSEC) {
+			bdp += 4;
+			bdp += bw >= 44738 ? 3 : 1;
+		} else if (minrtt >= 16 * USEC_PER_MSEC) {
+			bdp += 2;
+			bdp += bw >= 44738 ? 2 : 0;
+		} else if (minrtt >= 6 * USEC_PER_MSEC) {
+			bdp += 1;
+			bdp += bw >= 44738 ? 1 : 0;
+		}
+
 	return bdp;
 }
 
@@ -539,11 +576,19 @@
 
 	/* If we're below target cwnd, slow start cwnd toward target cwnd. */
 	if (bbr_full_bw_reached(sk))  /* only cut cwnd if we filled the pipe */
-		cwnd = min(cwnd + acked, target_cwnd);
+		cwnd = min(cwnd + acked + (acked >> 3), target_cwnd);
 	else if (cwnd < target_cwnd || tp->delivered < TCP_INIT_CWND)
-		cwnd = cwnd + acked;
+		cwnd = cwnd + acked + (acked >> 3);
 	cwnd = max(cwnd, bbr_cwnd_min_target);
 
+	if (unlikely(bbr->min_rtt_us <= 750)) {
+		cwnd = max_t(u32, cwnd, 50);
+		if (bbr->mode == BBR_PROBE_RTT)
+			cwnd = max_t(u32, cwnd / 2, 24);
+		tcp_snd_cwnd_set(tp, min(cwnd, tp->snd_cwnd_clamp));
+		return;
+	}
+
 done:
 	tcp_snd_cwnd_set(tp, min(cwnd, tp->snd_cwnd_clamp));	/* apply global cap */
 	if (bbr->mode == BBR_PROBE_RTT)  /* drain queue, refresh min_rtt */
@@ -987,11 +1032,12 @@
 static void bbr_update_gains(struct sock *sk)
 {
 	struct bbr *bbr = inet_csk_ca(sk);
+	struct tcp_sock *tp = tcp_sk(sk);
 
 	switch (bbr->mode) {
 	case BBR_STARTUP:
-		bbr->pacing_gain = bbr_high_gain;
-		bbr->cwnd_gain	 = bbr_high_gain;
+		bbr->pacing_gain = tp->app_limited ? bbr_limited_pacing_gain : bbr_high_gain;
+		bbr->cwnd_gain	 = bbr_startup_cwnd_gain;
 		break;
 	case BBR_DRAIN:
 		bbr->pacing_gain = bbr_drain_gain;	/* slow, to drain */
