--- a/kernel/power/energy_model.c	2026-04-03 16:06:32.382147711 +0800
+++ b/kernel/power/energy_model.c	2026-04-04 02:28:08.214058362 +0800
@@ -281,8 +281,8 @@
 		table[i].cost = cost;
 
 		if (table[i].cost >= prev_cost) {
-			table[i].flags = EM_PERF_STATE_INEFFICIENT;
-			dev_dbg(dev, "EM: OPP:%lu is inefficient\n",
+			//table[i].flags = EM_PERF_STATE_INEFFICIENT;
+			dev_dbg(dev, "EM: OPP:%lu is inefficient, but is ignored!\n",
 				table[i].frequency);
 		} else {
 			prev_cost = table[i].cost;
